lib/tests/streamlit/elements/element_utils_test.py

```diff
@@ -147,3 +154,85 @@ def test_check_session_state_rules_writes_not_allowed(
+            C = enum.auto()
+
+        EnumAOrig = EnumA
+
+        class EnumA(enum.Enum):
+            A = enum.auto()
+            B = enum.auto()
+            C = enum.auto()
+
+        EnumAEqual = EnumA
+        EnumAEqualList = [EnumAEqual.A, EnumAEqual.C, EnumAEqual.B]
+
+        int_result = RegisterWidgetResult(1, False)
+        intlist_result = RegisterWidgetResult([1, 2, 3], False)
+
+        single_result = RegisterWidgetResult(EnumAOrig.A, False)
+        single_coerced = RegisterWidgetResult(EnumAEqual.A, False)
+
+        tuple_result = RegisterWidgetResult((EnumAOrig.A, EnumAOrig.C), True)
+        tuple_coerced = RegisterWidgetResult((EnumAEqual.A, EnumAEqual.C), True)
+
+        list_result = RegisterWidgetResult([EnumAOrig.A, EnumAOrig.C], True)
+        list_coerced = RegisterWidgetResult([EnumAEqual.A, EnumAEqual.C], True)
+
+        assert maybe_coerce_enum(single_result, EnumAEqual, []) == single_coerced
+        assert (
+            maybe_coerce_enum(single_result, EnumAEqualList, EnumAEqualList)
+            == single_coerced
+        )
+        assert (
+            maybe_coerce_enum(single_result, EnumAEqualList, [EnumAEqual.A])
+            == single_coerced
+        )
+        assert maybe_coerce_enum(single_result, [1, 2, 3], []) is single_result
+        assert maybe_coerce_enum(int_result, EnumAEqual, []) is int_result
+        assert (
+            maybe_coerce_enum(
+                single_result, EnumAEqualList, [EnumAEqual.A, EnumAOrig.B]
+            )
+            is single_result
+        )
+
+        assert maybe_coerce_enum_sequence(tuple_result, EnumAEqual, []) == tuple_coerced
+        assert (
+            maybe_coerce_enum_sequence(tuple_result, EnumAEqualList, EnumAEqualList)
+            == tuple_coerced
+        )
+        assert (
+            maybe_coerce_enum_sequence(tuple_result, EnumAEqualList, [EnumAEqual.A])
+            == tuple_coerced
+        )
+        assert maybe_coerce_enum_sequence(list_result, EnumAEqual, []) == list_coerced
+        assert (
+            maybe_coerce_enum_sequence(list_result, EnumAEqualList, EnumAEqualList)
+            == list_coerced
+        )
+        assert (
+            maybe_coerce_enum_sequence(list_result, EnumAEqualList, [EnumAEqual.A])
+            == list_coerced
+        )
+        assert maybe_coerce_enum_sequence(list_result, [1, 2, 3], []) is list_result
+        assert maybe_coerce_enum_sequence(tuple_result, [1, 2, 3], []) is tuple_result
+        assert (
+            maybe_coerce_enum_sequence(intlist_result, EnumAEqual, []) is intlist_result
+        )
+        assert (
+            maybe_coerce_enum_sequence(
+                list_result, EnumAEqualList, [EnumAEqual.A, EnumAOrig.B]
+            )
+            is list_result
+        )
+        assert (
+            maybe_coerce_enum_sequence(
+                tuple_result, EnumAEqualList, [EnumAEqual.A, EnumAOrig.B]
+            )
+            is tuple_result
+        )
```

Comment:
I think it'd be useful for readability to maybe do some common setup in a new test class so that the `assert`s themselves can be grouped into a few `test_` functions with names detailing what's going on in them (any reasonable grouping would be ok with me. The first that comes to mind would be to have a test for builtin types, one for singletons, etc.)